### PR TITLE
[CARBONDATA-3781] Refactor code to optimize partition pruning

### PR DESCRIPTION
**Pull request by Indhumathi27**
### Why is this PR needed?
If number of partitions is more on a segment, cost of checking if (indexPath  equalsIgnoreCase partitionsToPrune) in BlockletIndexFactory.getTableBlockUniqueIdentifierWrappers() is high.
 
 ### What changes were proposed in this PR?
1. Store location of partition filter in a SET and check if it contains indexPath.
2. Remove unused variables
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1022/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2734/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1023/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2735/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2737/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1025/

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletIndexFactory.java [hunk 1]
```diff
@@ -193,21 +192,18 @@ public IndexBuilder createBuilder(Segment segment, String shardName,
...
         // get the indexFile Parent path and compare with the PartitionPath, if matches, then add
         // the corresponding tableBlockIndexUniqueIdentifier for pruning
-        for (PartitionSpec partitionSpec : partitionsToPrune) {
-          if (partitionSpec.getLocation().toString()
-              .equalsIgnoreCase(tableBlockIndexUniqueIdentifier.getIndexFilePath())) {
-            tableBlockIndexUniqueIdentifierWrappers.add(
-                new TableBlockIndexUniqueIdentifierWrapper(tableBlockIndexUniqueIdentifier,
-                    this.getCarbonTable()));
-          }
+        if (partitionsToPrune.contains(tableBlockIndexUniqueIdentifier.getIndexFilePath())) {
```
Comment (marchpure) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletIndexFactory.java:203]: I tested failed.
In this dli spark, the indexfilepath is "obs://bucktable//tablename//partitionname",
while the partitionSpec's location is "obs://bucktable/tablename/partitionname"
Leading to pruning empty blocklets when queried partition table

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2763/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2766/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1053/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1055/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2768/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1056/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2769/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2771/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1058/

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1091/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2803/

File: core/src/main/java/org/apache/carbondata/core/index/TableIndex.java [hunk 2]
```diff
@@ -187,8 +191,18 @@ public CarbonTable getTable() {
     return blocklets;
   }
 
+  private Set<Path> getPartitionsToPrune(List<PartitionSpec> partitionSpecs) {
```
Comment (akashrn5) [inline core/src/main/java/org/apache/carbondata/core/index/TableIndex.java:194]: i think its better to rename the method name as `getPartionLocations`

**Comment by akashrn5** [comment]
@Indhumathi27 As we are checking the contains for the index Path and partition localtion, its better to say at place, instead of `Set<Path> partitions` , we can say `Set<Path> partitionLocaltions`. This is more better i think.

**Comment by akashrn5** [comment]
@Indhumathi27 , this is an improvement, can you please add a jira and change the PR heading?

File: core/src/main/java/org/apache/carbondata/core/index/TableIndex.java (hunk 2 above)
Comment (Indhumathi27) [reply core/src/main/java/org/apache/carbondata/core/index/TableIndex.java]: done

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2820/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1107/

**Comment by Indhumathi27** [comment]
retest this please

File: core/src/main/java/org/apache/carbondata/core/index/TableIndex.java [hunk 3]
```diff
@@ -177,18 +181,29 @@ public CarbonTable getTable() {
...
+      for (PartitionSpec partitionSpec : partitionSpecs) {
+        partitionsLocations.add(partitionSpec.getLocation());
+      }
+    }
+    return partitionsLocations;
+  }
+
   private List<ExtendedBlocklet> pruneWithFilter(List<Segment> segments, IndexFilter filter,
-      List<PartitionSpec> partitions, List<ExtendedBlocklet> blocklets,
+      Set<Path> partitionLocations, List<ExtendedBlocklet> blocklets,
```
Comment (kunal642) [inline core/src/main/java/org/apache/carbondata/core/index/TableIndex.java:206]: Why not make Set<PartitionSpec> ??

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2842/

File: core/src/main/java/org/apache/carbondata/core/index/TableIndex.java (hunk 3 above)
Comment (Indhumathi27) [reply core/src/main/java/org/apache/carbondata/core/index/TableIndex.java:206]: Only Partition Location is used from PartiitonSpec. whole PartiitonSpec object is not needed

**Comment by Indhumathi27** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1130/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1136/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2848/

**Comment by akashrn5** [comment]
LGTM

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**